threeML/plugins/SpectrumLike.py

```diff
@@ -15,26 +15,27 @@
 
 from threeML.config.config import threeML_config
 from threeML.exceptions.custom_exceptions import custom_warnings, NegativeBackground
-import threeML.io.plotting.data_residual_plot
 from threeML.io.plotting.light_curve_plots import channel_plot, disjoint_patch_plot
 from threeML.io.rich_display import display
 from threeML.plugin_prototype import PluginPrototype
 from threeML.plugins.XYLike import XYLike
-from threeML.utils.OGIP.likelihood_functions import half_chi2
-from threeML.utils.OGIP.likelihood_functions import poisson_log_likelihood_ideal_bkg
-from threeML.utils.OGIP.likelihood_functions import poisson_observed_gaussian_background
-from threeML.utils.OGIP.likelihood_functions import poisson_observed_poisson_background
 from threeML.utils.binner import Rebinner
 from threeML.utils.spectrum.binned_spectrum import BinnedSpectrum, ChannelSet
-from threeML.utils.stats_tools import Significance
+
 from threeML.utils.string_utils import dash_separated_string_to_tuple
+from threeML.utils.spectrum.pha_spectrum import PHASpectrum
+
+from threeML.utils.statistics.stats_tools import Significance
+from threeML.utils.spectrum.spectrum_likelihood import likelihood_lookup
+from threeML.io.plotting.data_residual_plot import ResidualPlot
+
 
 NO_REBIN = 1E-99
 
 __instrument_name = "General binned spectral data"
 
 # This defines the known noise models for source and/or background spectra
-_known_noise_models = ['poisson', 'gaussian', 'ideal']
+_known_noise_models = ['poisson', 'gaussian', 'ideal', 'modeled']
 
 
 class SpectrumLike(PluginPrototype):
```

Comment:
The constructor is quite long with a lot of `if/else` and Codecov does not like it. I suggest factoring it in multiple shorter methods. For example, lines 93-230 can be placed in a method like  `_probe_noise_models`. If you decide to go down this way, please do not assign members of the class from within the new method, but return them instead. For example:

```python
# WRONG
def __init__(self):
    self._method()

def _method(self):
    self._member1 = 'foo'
    self._member2 = 'bar'

# CORRECT
def __init__(self):
    self._member1, self._member2 = self._method()

def _method(self):
    member1 = 'foo'
    member2 = 'bar'
    return member1, member2

Reply:
I agree, we've been reducing it down overtime. It seems we keep coming back to this one :) I'll do some work on it and commit here.

Reply:
Um. Is the setting of class members outside of the constructor an issue with pickling? Can you point me to some guidelines? This becomes a little cumbersome in tis case because there are a lot of things set in nested logic.